robottelo/host_helpers/cli_factory.py

```diff
@@ -0,0 +1,285 @@
+        'root-password': gen_alpha,
+    },
+    'host_collection': {
+        'name': gen_alpha,
+    },
+    'job_invocation': {},
+    'job_template': {
+        'job-category': 'Miscellaneous',
+        'provider-type': 'SSH',
+    },
+    'user': {
+        '_setup': gen_alphanumeric,  # create a login name
+        'auth-source-id': 1,
+        'firstname': gen_alphanumeric,
+        'lastname': gen_alphanumeric,
+        'login': lambda: ENTITY_FIELDS['user']['_setup_res'],
+        'mail': lambda: f"{ENTITY_FIELDS['user']['_setup_res']}@example.com",
+        'password': gen_alphanumeric,
+    },
+    'usergroup': {
+        'name': gen_alphanumeric,
+    },
+    'usergroup_external': {
+        'auth-source-id': 1,
+        'name': gen_alphanumeric,
+    },
+    'ldap_auth_source': {
+        'name': gen_alphanumeric,
+    },
+    'compute_resource': {
+        'name': gen_alphanumeric,
+    },
+    'org': {'_redirect': 'org_with_credentials'},
+    'org_with_credentials': {
+        '_entity_cls': 'Org',
+        'name': gen_alphanumeric,
+    },
+    'realm': {
+        'name': gen_alphanumeric,
+    },
+    'report_template': {
+        'file': gen_alphanumeric,  # formerly content
+        'name': gen_alphanumeric,
+    },
+    'os': {
+        'major': partial(random.randint, 0, 10),
+        'minor': partial(random.randint, 0, 10),
+        'name': gen_alphanumeric,
+    },
+    'scapcontent': {
+        '_entity_cls': 'Scapcontent',
+        'title': gen_alphanumeric().lower(),
+    },
+    'domain': {
+        '_entity_cls': 'Domain',
+        'name': lambda: gen_alphanumeric().lower(),
+    },
+    'hostgroup': {
+        '_entity_cls': 'HostGroup',
+        'name': gen_alphanumeric,
+    },
+    'medium': {
+        '_entity_cls': 'Medium',
+        'name': gen_alphanumeric,
+        'path': lambda: f'http://{gen_alpha()}',
+    },
+    'environment': {
+        '_entity_cls': 'Environment',
+        'name': gen_alphanumeric,
+    },
+    'lifecycle_environment': {
+        '_entity_cls': 'LifecycleEnvironment',
+        'name': gen_alphanumeric,
+    },
+    'tailoringfile': {
+        '_entity_cls': 'TailoringFiles',
+        'name': lambda: gen_alphanumeric().lower(),
+    },
+    'template': {
+        'file': lambda: f'/root/{gen_alphanumeric()}',
+        'name': gen_alphanumeric,
+        'type': partial(random.choice, constants.TEMPLATE_TYPES),
+    },
+    'template_input': {
+        '_entity_cls': 'TemplateInput',
+        'name': gen_alphanumeric,
+    },
+    'virt_who_config': {
+        'filtering-mode': 'none',
+        'hypervisor-id': 'hostname',
+        'interval': '60',
+        'name': gen_alphanumeric,
+        'satellite-url': lambda self: self._satellite.hostname,
+    },
+}
+
+
+class CLIFactory:
+    def __init__(self, satellite):
+        self._satellite = satellite
+
+    def __getattribute__(self, name):
+        """Handle cases when a make_ method is called (e.g. make_activation_key)"""
```

Comment:
Please add a lot more in-line documentation of the steps you're taking here and the intent of the functions in this class.

Reply:
If you have any specific areas of confusion, that would help me figure out all of what needs to be said. I don't want to explain what might be obvious.

Reply:
See if the additional comments help more. I don't want to come anywhere close to the overkill that is nailgun's documentation or teach what `__getattribute__` does via comments. However, the existing hints should give enough information to explain what it going on.

Reply:
The updated docblock (and I think some in-line comments are new?) help explain the purpose. Certainly don't need to document when/why/how `__getattribute__` itself is triggered, just why/how we're overriding it and what actions we're taking on the args or what's being called.

I think some more in-line documentation surrounding the field function execution would help. Placed above the dict definition, so when reading its not a guess why this giant dict is defined with some function objects being set as values but not evaluated.